### DISCUSSION TITLE
How can my SyncManager pull deletes?

### DISCUSSION THREAD

**Discussion by MechMel** [maxnowack/signaldb#1046] [Q&A] [answered]
I'm tying to make a firestore SyncManager, and I'm not sure how Signal DB expects me to tell it when docs are deleted. Most other local first tools use tombstones, but SignalDB seems to be performing actual deletes. I also don't see a way to emit delete events in the SyncManager's pull function. What is the right way to approach this?

```ts
import { SyncManager } from 'signaldb'
import { initializeApp } from 'firebase/app'
import { getDatabase, ref, get, set, remove, update, onChildAdded, onChildChanged, onChildRemoved } from 'firebase/database'

initializeApp({
  databaseURL: 'https://signaldb-d7e71-default-rtdb.firebaseio.com',
})
const db = getDatabase()

const syncManager = new SyncManager({
  onError: (options, error) => {
    // eslint-disable-next-line no-console
    console.error(options, error)
  },
  registerRemoteChange({ name }, onChange) {
    const handleChange = () => {
      void onChange()
    }
    onChildAdded(ref(db, name), () => { void handleChange() })
    onChildChanged(ref(db, name), () => { void handleChange() })
    onChildRemoved(ref(db, name), () => { void handleChange() })
  },
  async pull({ name }) {
    const snapshot = await get(ref(db, name))
    const items = await snapshot.val() as Record<string, any> | null
    return { items: Object.values(items ?? {}) }
  },
  async push({ name }, { changes }) {
    await Promise.all([
      ...changes.added.map(async (item) => {
        await set(ref(db, `${name}/${item.id}`), item)
      }),
      ...changes.modified.map(async (item) => {
        await update(ref(db, `${name}/${item.id}`), item)
      }),
      ...changes.removed.map(async (item) => {
        await remove(ref(db, `${name}/${item.id}`))
      }),
    ])
  },
})

export default syncManager
```

**Comment by maxnowack** [comment, marked as answer]
If the `pull` method returns an object with the `items` property, the content of the SignalDB collection will be replaced with this data. This means, if the response doesn't contain a specific item. It will be deleted locally. You could also return an object with a `changes` property, if you want to control this by yourself.

> The `pull` method must return a promise that resolves to an object with either an `items` property containing all items that should be applied to the collection or a `changes` property containing all changes `{ added: T[], modified: T[], removed: T[] }`.
>
> *from https://signaldb.js.org/sync/implementation/#implementing-the-pull-method*

**Reply by SemAiSoft** [reply]
`{ added: T[], modified: T[], removed: T[] }`
when returned as such, is still deleting the entire local data.

**Reply by maxnowack** [reply]
@SemAiSoft could you provide a minimal reproduction for this?
